File: Longest Bitonic subsequence.cpp
class Solution{
	public:
	int LongestBitonicSequence(vector<int>nums)
	{
	    vector<int> Incre(nums.size(), 0 ), Dec(nums.size(), 0);
	    
	    Incre[0] = 1;
	    for(int i=1; i<nums.size(); i++){
	        int mx = 0;
	        for(int j=0;j<i;j++){
	            if(nums[j]<nums[i])
	                mx = max(mx, Incre[j]);
	        }
	        Incre[i] = mx + 1;
	    }
	    
	    Dec[nums.size()-1] = 1;
	    for(int i=nums.size()-2; i>=0; i--){
	        int mx = 0;
	        for(int j=i+1;j<nums.size();j++){
	            if(nums[j]<nums[i])
	                mx = max(mx, Dec[j]);
	        }
	        Dec[i] = mx + 1;
	    }
	    
	    int ans = 0;
	    
	    for(int i=0; i<nums.size(); i++){
	        ans = max(ans, Incre[i] + Dec[i] - 1);
	    }
	    
	    return ans;
	}
};

File: GFG/Longest Bitonic subsequence/Longest Bitonic subsequence.java
class Solution
{
    public int LongestBitonicSequence(int[] nums)
    {
        // Code here
        int n=nums.length;
        int[]inc=new int[n];
        for(int i=0;i<n;i++){
            inc[i]=1;
            for(int j=i-1;j>=0;j--){
                if(nums[j]<nums[i] && inc[j]+1>inc[i]){
                    inc[i]=inc[j]+1;
                }
            }
        }
        int[]dec=new int[n];
        for(int i=n-1;i>=0;i--){
            dec[i]=1;
            for(int j=i+1;j<n;j++){
                if(nums[j]<nums[i] && dec[j]+1>dec[i]){
                    dec[i]=dec[j]+1;
                }
            }
        }
        
        int ans=0;
        for(int i=0;i<n;i++){
            ans=Math.max(ans,inc[i]+dec[i]-1);
        }
        return ans;
    }
}
